File: Python/169. Majority Element..py
# 169. Majority Element.          -Easy-

# Topic: Array, Hash Table, Divide and Conquer, Sorting, Counting.

"""
### Task:
---
Given an array nums of size n, return the majority element.

The majority element is the element that appears more than ⌊n / 2⌋ times. You may assume that the majority 
element always exists in the array.

Example 1:
Input: nums = [3,2,3]
Output: 3

Example 2:
Input: nums = [2,2,1,1,1,2,2]
Output: 2

Constraints:
n == nums.length
1 <= n <= 5 * 10^4
-10^9 <= nums[i] <= 10^9

Follow-up: Could you solve the problem in linear time and in O(1) space?


### Testcase:
---
[3,2,3]
[2,2,1,1,1,2,2]


### Code:
---
class Solution:
    def majorityElement(self, nums: List[int]) -> int:

"""
### Solution: -------------------------------------

class Solution:
    def majorityElement(self, nums):
        # Initialize the candidate for the majority element and a counter
        candidate, count = None, 0

        for num in nums:
            # If the count is zero, we select the current element as our new candidate
            if count == 0:
                candidate = num
            # If the current element is the candidate, increment the count
            # Otherwise, decrement the count
            count += (1 if num == candidate else -1)

        # The candidate at the end is the majority element
        return candidate

# Testing the function
sol = Solution()
test_case_1 = [3, 2, 3]
test_case_2 = [2, 2, 1, 1, 1, 2, 2]

output_1 = sol.majorityElement(test_case_1)
output_2 = sol.majorityElement(test_case_2)

output_1, output_2


### Description: ===================================
'''

The "Majority Element" problem asks to find the element in an array that appears more than ⌊n / 2⌋ times, 
where n is the size of the array. A popular and efficient way to solve this problem in linear time and with 
constant space is using the Boyer-Moore Voting Algorithm.

### Boyer-Moore Voting Algorithm:
1. **Initialization**: Start with an initial candidate and a counter set to 0.
2. **Iterate Through the Array**:
   - If the counter is 0, choose the current element as the new candidate.
   - If the current element is the same as the candidate, increment the counter.
   - Otherwise, decrement the counter.
3. **Result**: The candidate at the end of the array is the majority element. This works because the majority element 
   is guaranteed to have more occurrences than all other elements combined.


### Explanation:
- The algorithm maintains a candidate for the majority element and a counter.
- Each time it finds an element equal to the candidate, it increments the counter.
- If it finds an element different from the candidate, it decrements the counter.
- When the counter reaches 0, it picks a new candidate from the current position in the array.
- Since the majority element appears more than ⌊n / 2⌋ times, it will always be the final candidate.

This approach has a time complexity of O(n) and a space complexity of O(1), fulfilling the follow-up challenge.

Let's execute the code to test the provided examples.

The implemented solution correctly identifies the majority element for the given test cases:

1. For the input `[3, 2, 3]`, the output is `3`. This is correct as `3` appears more than ⌊3 / 2⌋ = 1 times in the array.

2. For the input `[2, 2, 1, 1, 1, 2, 2]`, the output is `2`. This is also correct as `2` appears more than ⌊7 / 2⌋ = 3 times in the array.


### Explanation of Comments:
- **Initialization**: We start with no candidate and a counter set to 0.
- **Choosing a New Candidate**: When the count becomes zero, it means the previous candidate is not the majority. 
    Hence, we choose the current element as the new candidate.
- **Updating the Count**: If the current element is the same as our candidate, we increase the count as it's likely 
    to be the majority. If it's different, we decrease the count.
- **Final Candidate**: The algorithm ensures that the final candidate is the majority element due to its frequency in the array. 

This approach ensures that the majority element, which appears more than half the time in the array, is always selected as the 
final candidate. The comments added provide clarity on how the Boyer-Moore Voting Algorithm works in this context.

The Boyer-Moore Voting Algorithm efficiently finds the majority element in O(n) time and O(1) space, satisfying 
the problem's constraints and the follow-up challenge.

'''

File: Cpp/169. Majority Element.cpp
// 169. Majority Element.


// Topic: Array, Hash Table, Divide and Conquer, Sorting, Counting.


/*
### Task:
---
Given an array nums of size n, return the majority element.

The majority element is the element that appears more than ⌊n / 2⌋ times. You may assume that the majority element always exists in the array.

Example 1:
Input: nums = [3,2,3]
Output: 3

Example 2:
Input: nums = [2,2,1,1,1,2,2]
Output: 2

Constraints:
n == nums.length
1 <= n <= 5 * 104
-10^9 <= nums[i] <= 10^9
 
Follow-up: Could you solve the problem in linear time and in O(1) space?


### Testcase:
---
[3,2,3]
[2,2,1,1,1,2,2]


### Code:
---
class Solution {
public:
    int majorityElement(vector<int>& nums) {
        
    }
};

*/
// Solution: --------------------------------------

class Solution {
public:
    int majorityElement(vector<int>& nums) {
        // Initialize count and candidate for majority element
        int count = 0;
        int major;

        // Iterate through the array
        for (int num : nums) {
            // If count is 0, we choose the current element as the new candidate
            if (count == 0) {
                major = num;
            }

            // If the current element is the same as our current candidate, increase the count
            // This means the current candidate is gaining support
            if (num == major) {
                count++;
            } else {
                // If the current element is different, decrease the count
                // This means the current candidate is losing support
                count--;
            }
        }

        // The candidate remaining after the entire array traversal is the majority element
        return major;
    }
};




// Description: ===================================
/*

To solve the "Majority Element" problem, we can use the Boyer-Moore Voting Algorithm, which is efficient in terms of time 
(linear time complexity) and space (constant space complexity). The essence of this algorithm is to find a candidate for 
the majority element and then verify if it is indeed the majority element.

Here's how the Boyer-Moore Voting Algorithm works:
1. Initialize a candidate element `major` and a count `count` to 0.
2. Traverse through the array `nums`.
   - If `count` is 0, set `major` to the current element.
   - If the current element is the same as `major`, increment `count`.
   - Otherwise, decrement `count`.
3. Since the majority element appears more than ⌊n / 2⌋ times, the remaining element in `major` after the entire array is traversed 
   is the majority element.

### Description:
- We initialize `count` to 0 and `major` to an undefined value.
- As we iterate through `nums`, we set `major` to the current element `num` if `count` is 0. This is like saying, "Let's consider this new element as a candidate for the majority element."
- If the current element `num` is the same as our candidate `major`, we increment `count`, indicating that we've found another instance of our candidate.
- If `num` is different from `major`, we decrement `count`, indicating that there's a conflict, and our candidate might not be the majority element.
- By the end of the array, `major` holds the element that has survived these conflicts, which must be the majority element due to the problem's constraints.

This solution is efficient because it only requires a single pass through the array (`O(n)` time complexity) and uses only constant 
extra space (`O(1)` space complexity), meeting the follow-up challenge of solving the problem in linear time and constant space.

*/